src/plugins/data/public/search/aggs/buckets/terms_fn.ts

```diff
@@ -0,0 +1,126 @@
+/*
+ * Licensed to Elasticsearch B.V. under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch B.V. licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+import { i18n } from '@kbn/i18n';
+import { ExpressionFunctionDefinition } from '../../../../../expressions/public';
+import { AggExpressionType, AggExpressionFunctionArgs } from '../';
+
+const aggTypeName = 'terms';
+const fnName = 'aggTypeTerms';
+
+type Input = any;
+type Arguments = AggExpressionFunctionArgs<typeof aggTypeName>;
+type Output = AggExpressionType;
+type FunctionDefinition = ExpressionFunctionDefinition<typeof fnName, Input, Arguments, Output>;
+
+export const aggTypeTerms = (): FunctionDefinition => ({
+  name: fnName,
+  help: i18n.translate('data.search.aggs.function.buckets.terms.help', {
+    defaultMessage: 'Generates AggConfigJSON for a terms agg',
+  }),
+  type: 'agg_type',
+  args: {
+    id: {
+      types: ['string'],
+      help: '',
+    },
+    enabled: {
+      types: ['boolean'],
+      default: true,
+      help: '',
+    },
+    schema: {
+      types: ['string'],
+      help: '',
+    },
+    field: {
+      types: ['string'],
+      required: true,
+      help: i18n.translate('data.search.aggs.buckets.terms.field.help', {
+        defaultMessage: 'Field to use for terms aggregation',
+      }),
+    },
+    order: {
+      types: ['string'],
+      required: true,
+      help: '',
+    },
+    orderBy: {
+      types: ['string'],
+      help: '',
+    },
+    orderAgg: {
+      types: ['agg_type'],
```

Comment:
order/orderBy/orderAgg .... thats definetely confusing.
is this something that is used in other aggregations ?

Reply:
and are orderAgg and orderBy exclusive ? either orderBy an already define aggregation or provide a new one in orderAgg ? 

if that's the case we could:
1. extract both into a new function, which always returns an aggregation 
2. remove orderBy argument, make sure build_pipeline finds the correct aggregation by id and generates expressioin from it and passes it as orderAgg 
3. remove orderAgg argument, make sure build_pipeline adds a new aggregation to esaggs and uses it id as orderBy

but as this seems to be used only in terms agg i am fine with leaving it as it is for now. but if we would go with 3. that would mean terms_fn doesn't have any arguments accepting complex types.

Reply:
I agree it's confusing having these 3 for terms aggs, however i did manage to get it working using the subexpression. In the end you should still get back the serialized agg for the param value.

So I'd lean toward revisiting this later rather than adding further refactoring here, but I'm open to discussing that further.